cparser/Elab.ml

```diff
@@ -1644,27 +1690,27 @@ let elab_expr vararg loc env a =
                 err "illegal pointer arithmetic in binary '-'";
               (TPtr(ty, []), TPtr(ty, []))
           | (TInt _ | TEnum _), (TPtr(ty, a) | TArray(ty, _, a)) ->
-              if not (wrap pointer_arithmetic_ok loc env ty) then
-                err "illegal pointer arithmetic in binary '-'";
+              check_ptr_arith env ty;
               (TPtr(ty, []), TPtr(ty, []))
           | (TPtr(ty1, a1) | TArray(ty1, _, a1)),
             (TPtr(ty2, a2) | TArray(ty2, _, a2)) ->
               if not (compatible_types AttrIgnoreAll env ty1 ty2) then
-                err "mismatch between pointer types in binary '-'";
-              if not (pointer_arithmetic_ok env ty1) then
-                err "illegal pointer arithmetic in binary '-'";
+                err "%a and %a are not pointers to compatible types"
+                   (print_typ env) b1.etyp (print_typ env) b1.etyp;
+              check_ptr_arith env ty1;
               if wrap sizeof loc env ty1 = Some 0 then
                 err "subtraction between two pointers to zero-sized objects";
               (TPtr(ty1, []), TInt(ptrdiff_t_ikind(), []))
-          | _, _ -> error "type error in binary '-'"
+          | _, _ -> error "invalid operands to binary expression (%a and %a)"
+                (print_typ env) b1.etyp (print_typ env) b2.etyp
         end in
       { edesc = EBinop(Osub, b1, b2, tyop); etyp = tyres },env
 
   | BINARY(SHL, a1, a2) ->
-      elab_shift "<<" Oshl a1 a2
+      elab_binary_integer Oshl a1 a2
```

Comment:
NOOOOOO!  The typing rules for shifts are different from those for binary arithmetic operators!  Did you read the original code before hacking it???